man/mysql.1
'\" t
.\"     Title: \fBmysql\fR
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 02/04/2010
.\"    Manual: MySQL Database System
.\"    Source: MySQL 5.1
.\"  Language: English
.\"
.TH "\FBMYSQL\FR" "1" "02/04/2010" "MySQL 5\&.1" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.\" mysql
.\" command-line tool
.\" tools: command-line
.\" scripts: SQL
.\" SQL scripts
.\" batch SQL files
.SH "NAME"
mysql \- the MySQL command\-line tool
.SH "SYNOPSIS"
.HP \w'\fBmysql\ [\fR\fB\fIoptions\fR\fR\fB]\ \fR\fB\fIdb_name\fR\fR\ 'u
\fBmysql [\fR\fB\fIoptions\fR\fR\fB] \fR\fB\fIdb_name\fR\fR
.SH "DESCRIPTION"
.PP
\fBmysql\fR
is a simple SQL shell (with GNU
readline
capabilities)\&. It supports interactive and noninteractive use\&. When used interactively, query results are presented in an ASCII\-table format\&. When used noninteractively (for example, as a filter), the result is presented in tab\-separated format\&. The output format can be changed using command options\&.
.PP
If you have problems due to insufficient memory for large result sets, use the
\fB\-\-quick\fR
option\&. This forces
\fBmysql\fR
to retrieve results from the server a row at a time rather than retrieving the entire result set and buffering it in memory before displaying it\&. This is done by returning the result set using the
mysql_use_result()
C API function in the client/server library rather than
mysql_store_result()\&.
.PP
Using
\fBmysql\fR
is very easy\&. Invoke it from the prompt of your command interpreter as follows:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBmysql \fR\fB\fIdb_name\fR\fR
.fi
.if n \{\
.RE
.\}
.PP
Or:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBmysql \-\-user=\fR\fB\fIuser_name\fR\fR\fB \-\-password=\fR\fB\fIyour_password\fR\fR\fB \fR\fB\fIdb_name\fR\fR
.fi
.if n \{\
.RE
.\}
.PP
Then type an SQL statement, end it with
\(lq;\(rq,
\eg, or
\eG
and press Enter\&.
.PP
As of MySQL 5\&.1\&.10, typing Control\-C causes
\fBmysql\fR
to attempt to kill the current statement\&. If this cannot be done, or Control\-C is typed again before the statement is killed,
\fBmysql\fR
exits\&. Previously, Control\-C caused
\fBmysql\fR
to exit in all cases\&.
.PP
You can execute SQL statements in a script file (batch file) like this:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBmysql \fR\fB\fIdb_name\fR\fR\fB < \fR\fB\fIscript\&.sql\fR\fR\fB > \fR\fB\fIoutput\&.tab\fR\fR
.fi
.if n \{\
.RE
.\}
.SH "MYSQL OPTIONS"
.\" mysql command options
.\" command options: mysql
.\" options: command-line: mysql
.\" startup parameters: mysql
.PP
\fBmysql\fR
supports the following options, which can be specified on the command line or in the
[mysql]
and
[client]
option file groups\&.
\fBmysql\fR
also supports the options for processing option files described at
Section\ \&4.2.3.3.1, \(lqCommand-Line Options that Affect Option-File Handling\(rq\&.
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: help option
.\" help option: mysql
\fB\-\-help\fR,
\fB\-?\fR
.sp
Display a help message and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: auto-rehash option
.\" auto-rehash option: mysql
\fB\-\-auto\-rehash\fR
.sp
Enable automatic rehashing\&. This option is on by default, which enables database, table, and column name completion\&. Use
\fB\-\-disable\-auto\-rehash\fR
to disable rehashing\&. That causes
\fBmysql\fR
to start faster, but you must issue the
rehash
command if you want to use name completion\&.
.sp
To complete a name, enter the first part and press Tab\&. If the name is unambiguous,
\fBmysql\fR
completes it\&. Otherwise, you can press Tab again to see the possible names that begin with what you have typed so far\&. Completion does not occur if there is no default database\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: batch option
.\" batch option: mysql
\fB\-\-batch\fR,
\fB\-B\fR
.sp
Print results using tab as the column separator, with each row on a new line\&. With this option,
\fBmysql\fR
does not use the history file\&.
.sp
Batch mode results in nontabular output format and escaping of special characters\&. Escaping may be disabled by using raw mode; see the description for the
\fB\-\-raw\fR
option\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: character-sets-dir option
.\" character-sets-dir option: mysql
\fB\-\-character\-sets\-dir=\fR\fB\fIpath\fR\fR
.sp
The directory where character sets are installed\&. See
Section\ \&9.2, \(lqThe Character Set Used for Data and Sorting\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: column-names option
.\" column-names option: mysql
\fB\-\-column\-names\fR
.sp
Write column names in results\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: column-type-info option
.\" column-type-info option: mysql
\fB\-\-column\-type\-info\fR,
\fB\-m\fR
.sp
Display result set metadata\&. This option was added in MySQL 5\&.1\&.14\&. (Before that, use
\fB\-\-debug\-info\fR\&.) The
\fB\-m\fR
short option was added in MySQL 5\&.1\&.21\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: comments option
.\" comments option: mysql
\fB\-\-comments\fR,
\fB\-c\fR
.sp
Whether to preserve comments in statements sent to the server\&. The default is \-\-skip\-comments (discard comments), enable with \-\-comments (preserve comments)\&. This option was added in MySQL 5\&.1\&.23\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: compress option
.\" compress option: mysql
\fB\-\-compress\fR,
\fB\-C\fR
.sp
Compress all information sent between the client and the server if both support compression\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: database option
.\" database option: mysql
\fB\-\-database=\fR\fB\fIdb_name\fR\fR,
\fB\-D \fR\fB\fIdb_name\fR\fR
.sp
The database to use\&. This is useful primarily in an option file\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: debug option
.\" debug option: mysql
\fB\-\-debug[=\fR\fB\fIdebug_options\fR\fR\fB]\fR,
\fB\-# [\fR\fB\fIdebug_options\fR\fR\fB]\fR
.sp
Write a debugging log\&. A typical
\fIdebug_options\fR
string is
\'d:t:o,\fIfile_name\fR\'\&. The default is
\'d:t:o,/tmp/mysql\&.trace\'\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: debug-check option
.\" debug-check option: mysql
\fB\-\-debug\-check\fR
.sp
Print some debugging information when the program exits\&. This option was added in MySQL 5\&.1\&.21\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: debug-info option
.\" debug-info option: mysql
\fB\-\-debug\-info\fR,
\fB\-T\fR
.sp
Before MySQL 5\&.1\&.14, this option prints debugging information and memory and CPU usage statistics when the program exits, and also causes display of result set metadata during execution\&. As of MySQL 5\&.1\&.14, use
\fB\-\-column\-type\-info\fR
to display result set metadata\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: default-character-set option
.\" default-character-set option: mysql
\fB\-\-default\-character\-set=\fR\fB\fIcharset_name\fR\fR
.sp
Use
\fIcharset_name\fR
as the default character set for the client and connection\&.
.sp
A common issue that can occur when the operating system uses
utf8
or another multi\-byte character set is that output from the
\fBmysql\fR
client is formatted incorrectly, due to the fact that the MySQL client uses the
latin1
character set by default\&. You can usually fix such issues by using this option to force the client to use the system character set instead\&.
.sp
See
Section\ \&9.2, \(lqThe Character Set Used for Data and Sorting\(rq, for more information\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: delimiter option
.\" delimiter option: mysql
\fB\-\-delimiter=\fR\fB\fIstr\fR\fR
.sp
Set the statement delimiter\&. The default is the semicolon character (\(lq;\(rq)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: disable named commands
.\" disable named command: mysql
\fB\-\-disable\-named\-commands\fR
.sp
Disable named commands\&. Use the
\e*
form only, or use named commands only at the beginning of a line ending with a semicolon (\(lq;\(rq)\&.
\fBmysql\fR
starts with this option
\fIenabled\fR
by default\&. However, even with this option, long\-format commands still work from the first line\&. See
the section called \(lqMYSQL COMMANDS\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: execute option
.\" execute option: mysql
\fB\-\-execute=\fR\fB\fIstatement\fR\fR,
\fB\-e \fR\fB\fIstatement\fR\fR
.sp
Execute the statement and quit\&. The default output format is like that produced with
\fB\-\-batch\fR\&. See
Section\ \&4.2.3.1, \(lqUsing Options on the Command Line\(rq, for some examples\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: force option
.\" force option: mysql
\fB\-\-force\fR,
\fB\-f\fR
.sp
Continue even if an SQL error occurs\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: host option
.\" host option: mysql
\fB\-\-host=\fR\fB\fIhost_name\fR\fR,
\fB\-h \fR\fB\fIhost_name\fR\fR
.sp
Connect to the MySQL server on the given host\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: html option
.\" html option: mysql
\fB\-\-html\fR,
\fB\-H\fR
.sp
Produce HTML output\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: ignore-spaces option
.\" ignore-spaces option: mysql
\fB\-\-ignore\-spaces\fR,
\fB\-i\fR
.sp
Ignore spaces after function names\&. The effect of this is described in the discussion for the
IGNORE_SPACE
SQL mode (see
Section\ \&5.1.8, \(lqServer SQL Modes\(rq)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: line-numbers option
.\" line-numbers option: mysql
\fB\-\-line\-numbers\fR
.sp
Write line numbers for errors\&. Disable this with
\fB\-\-skip\-line\-numbers\fR\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: local-infile option
.\" local-infile option: mysql
\fB\-\-local\-infile[={0|1}]\fR
.sp
Enable or disable
LOCAL
capability for
LOAD DATA INFILE\&. With no value, the option enables
LOCAL\&. The option may be given as
\fB\-\-local\-infile=0\fR
or
\fB\-\-local\-infile=1\fR
to explicitly disable or enable
LOCAL\&. Enabling
LOCAL
has no effect if the server does not also support it\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: named-commands option
.\" named-commands option: mysql
\fB\-\-named\-commands\fR,
\fB\-G\fR
.sp
Enable named
\fBmysql\fR
commands\&. Long\-format commands are allowed, not just short\-format commands\&. For example,
quit
and
\eq
both are recognized\&. Use
\fB\-\-skip\-named\-commands\fR
to disable named commands\&. See
the section called \(lqMYSQL COMMANDS\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: no-auto-rehash option
.\" no-auto-rehash option: mysql
\fB\-\-no\-auto\-rehash\fR,
\fB\-A\fR
.sp
This has the same effect as
\fB\-skip\-auto\-rehash\fR\&. See the description for
\fB\-\-auto\-rehash\fR\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: no-beep option
.\" no-beep option: mysql
\fB\-\-no\-beep\fR,
\fB\-b\fR
.sp
Do not beep when errors occur\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: no-named-commands option
.\" no-named-commands option: mysql
\fB\-\-no\-named\-commands\fR,
\fB\-g\fR
.sp
Deprecated, use
\fB\-\-disable\-named\-commands\fR
instead\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: no-pager option
.\" no-pager option: mysql
\fB\-\-no\-pager\fR
.sp
Deprecated form of
\fB\-\-skip\-pager\fR\&. See the
\fB\-\-pager\fR
option\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: no-tee option
.\" no-tee option: mysql
\fB\-\-no\-tee\fR
.sp
Do not copy output to a file\&.
the section called \(lqMYSQL COMMANDS\(rq, discusses tee files further\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: one-database option
.\" one-database option: mysql
\fB\-\-one\-database\fR,
\fB\-o\fR
.sp
Ignore statements except those for the default database named on the command line\&. This is useful for skipping updates to other databases in the binary log\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: pager option
.\" pager option: mysql
\fB\-\-pager[=\fR\fB\fIcommand\fR\fR\fB]\fR
.sp
Use the given command for paging query output\&. If the command is omitted, the default pager is the value of your
PAGER
environment variable\&. Valid pagers are
\fBless\fR,
\fBmore\fR,
\fBcat [> filename]\fR, and so forth\&. This option works only on Unix and only in interactive mode\&. To disable paging, use
\fB\-\-skip\-pager\fR\&.
the section called \(lqMYSQL COMMANDS\(rq, discusses output paging further\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: password option
.\" password option: mysql
\fB\-\-password[=\fR\fB\fIpassword\fR\fR\fB]\fR,
\fB\-p[\fR\fB\fIpassword\fR\fR\fB]\fR
.sp
The password to use when connecting to the server\&. If you use the short option form (\fB\-p\fR), you
\fIcannot\fR
have a space between the option and the password\&. If you omit the
\fIpassword\fR
value following the
\fB\-\-password\fR
or
\fB\-p\fR
option on the command line,
\fBmysql\fR
prompts for one\&.
.sp
Specifying a password on the command line should be considered insecure\&. See
Section\ \&5.5.6.2, \(lqEnd-User Guidelines for Password Security\(rq\&. You can use an option file to avoid giving the password on the command line\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: pipe option
.\" pipe option: mysql
\fB\-\-pipe\fR,
\fB\-W\fR
.sp
On Windows, connect to the server via a named pipe\&. This option applies only if the server supports named\-pipe connections\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: port option
.\" port option: mysql
\fB\-\-port=\fR\fB\fIport_num\fR\fR,
\fB\-P \fR\fB\fIport_num\fR\fR
.sp
The TCP/IP port number to use for the connection\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: prompt option
.\" prompt option: mysql
\fB\-\-prompt=\fR\fB\fIformat_str\fR\fR
.sp
Set the prompt to the specified format\&. The default is
mysql>\&. The special sequences that the prompt can contain are described in
the section called \(lqMYSQL COMMANDS\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: protocol option
.\" protocol option: mysql
\fB\-\-protocol={TCP|SOCKET|PIPE|MEMORY}\fR
.sp
The connection protocol to use for connecting to the server\&. It is useful when the other connection parameters normally would cause a protocol to be used other than the one you want\&. For details on the allowable values, see
Section\ \&4.2.2, \(lqConnecting to the MySQL Server\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
.\" mysql: quick option
.\" quick option: mysql
\fB\-\-quick\fR,
\fB\-q\fR
.sp

ChangeLog
------------------------------------------------------------
revno: 3193
tags: clone-5.1.41-build
committer: Georgi Kodinov <joro@sun.com>
branch nick: merge-5.1-bugteam
timestamp: Wed 2009-11-04 11:21:52 +0200
message:
  merge
------------------------------------------------------------
    revno: 1810.3961.16
    committer: Georgi Kodinov <joro@sun.com>
    branch nick: merge-5.0-bugteam
    timestamp: Wed 2009-11-04 11:13:55 +0200
message:
      version change
------------------------------------------------------------
revno: 3192
committer: Georgi Kodinov <joro@sun.com>
branch nick: merge-5.1-bugteam
timestamp: Wed 2009-11-04 11:18:34 +0200
message:
  merge
    ------------------------------------------------------------
    revno: 1810.3961.15
    committer: Georgi Kodinov <joro@sun.com>
    branch nick: merge-5.0-bugteam
    timestamp: Wed 2009-11-04 11:13:22 +0200
    message:
      Disabled the rpl_killed_ddl test in 5.0 because of bug #45520
------------------------------------------------------------
revno: 3191
committer: Davi Arnaut <Davi.Arnaut@Sun.COM>
branch nick: mysql-5.1-bugteam
timestamp: Tue 2009-11-03 16:18:44 -0200
message:
  Automerge.
    ------------------------------------------------------------
    revno: 1810.3961.14
    committer: Davi Arnaut <Davi.Arnaut@Sun.COM>
    branch nick: mysql-5.0-bugteam
    timestamp: Tue 2009-11-03 16:17:06 -0200
    message:
      Fix compiler warnings -- tag unused parameters.
    ------------------------------------------------------------
revno: 3190
committer: Konstantin Osipov <kostja@sun.com>
branch nick: 5.1-41756
timestamp: Tue 2009-11-03 20:45:52 +0300
    message:
  A fix and a test case for 
  Bug#41756 "Strange error messages about locks from InnoDB".
        
  In JT_EQ_REF (join_read_key()) access method, 
  don't try to unlock rows in the handler, unless certain that 
  a) they were locked
  b) they are not used.
  
  Unlocking of rows is done by the logic of the nested join loop,
  and is unaware of the possible caching that the access method may
  have. This could lead to double unlocking, when a row
  was unlocked first after reading into the cache, and then 
  when taken from cache, as well as to unlocking of rows which
  were actually used (but taken from cache).
        
  Delegate part of the unlocking logic to the access method,
  and in JT_EQ_REF count how many times a record was actually 
  used in the join. Unlock it only if it's usage count is 0.
  
  Implemented review comments.
        ------------------------------------------------------------
    revno: 1810.3961.13
    committer: Konstantin Osipov <kostja@sun.com>
    branch nick: 5.0-41756
    timestamp: Tue 2009-11-03 19:58:54 +0300
        message:
      A fix and a test case for
      Bug#41756 "Strange error messages about locks from InnoDB".
          
      In JT_EQ_REF (join_read_key()) access method,
      don't try to unlock rows in the handler, unless certain that
      a) they were locked
      b) they are not used.
          
      Unlocking of rows is done by the logic of the nested join loop,
      and is unaware of the possible caching that the access method may
      have. This could lead to double unlocking, when a row
      was unlocked first after reading into the cache, and then
      when taken from cache, as well as to unlocking of rows which
      were actually used (but taken from cache).
      
      Delegate part of the unlocking logic to the access method,
      and in JT_EQ_REF count how many times a record was actually
      used in the join. Unlock it only if it's usage count is 0.
      
      Implemented review comments.
    ------------------------------------------------------------
revno: 3189
committer: Magnus Bl?udd <magnus.blaudd@sun.com>
branch nick: 5.1-bugteam
timestamp: Tue 2009-11-03 18:07:19 +0100
    message:
  Merge bug#47867 to 5.1-bugteam
        ------------------------------------------------------------
    revno: 3146.2.1
    committer: Magnus Bl?udd <magnus.blaudd@sun.com>
    branch nick: 5.1-bug47867
    timestamp: Tue 2009-10-06 18:10:34 +0200
        message:
      Bug#47867 compiler warning _WIN32_WINNT macro redefinition
       - Remove the conflicting macro definitions since we define a higher value
         directly in the cmake files.
    ------------------------------------------------------------
revno: 3188
committer: Kristofer Pettersson <kristofer.pettersson@sun.com>
branch nick: mysql-5.1-bugteam
timestamp: Tue 2009-11-03 17:23:05 +0100
    message:
  automerge
        ------------------------------------------------------------
    revno: 3148.8.8
    committer: Sergey Vojtovich <svoj@sun.com>
    branch nick: mysql-5.1-bugteam-push
    timestamp: Tue 2009-11-03 18:44:39 +0400
        message:
      Merge of innodb-zip-ss6129 snapshot.
        ------------------------------------------------------------
        revno: 3148.13.37
        committer: Sergey Vojtovich <svoj@sun.com>
        branch nick: mysql-5.1-bugteam
        timestamp: Tue 2009-11-03 18:41:04 +0400
        message:
          Clean-up after applying innodb-zip-ss6129 snapshot:
          - re-enabled main.innodb_bug44369;
          - re-enabled main.innodb_bug47777;
          - re-enabled innodb.innodb_information_schema.
    ------------------------------------------------------------
        revno: 3148.13.36
        committer: Sergey Vojtovich <svoj@sun.com>
        branch nick: mysql-5.1-bugteam
        timestamp: Tue 2009-11-03 14:34:38 +0400
    message:
          Applying InnoDB plugin snashot 
            
          Detailed revision comments:
            
          r6126 | vasil | 2009-10-30 10:36:07 +0200 (Fri, 30 Oct 2009) | 45 lines
          branches/zip: Merge r6112:6125 from branches/5.1:
            
          (skipping r6122 and r6123, Jimmy says these are already present and need
          not be merged):
            

support-files/mysql.server-facebook.sh
#!/bin/bash
#
# mysqld	This shell script takes care of starting and stopping
#		the MySQL subsystem (mysqld).
#
# chkconfig: - 64 36
# description:	MySQL database server.
# processname: mysqld
# config: /etc/my.cnf
# pidfile: /var/run/mysqld/mysqld.pid

# Source function library.
. /etc/rc.d/init.d/functions

# Source networking configuration.
. /etc/sysconfig/network


prog="MySQL"

if [ "`echo $@ | grep -Pc '(\s+|^)--force(\s+|$)'`" == "1" ]; then
  force=1
else
  force=0
fi

# extract value of a MySQL option from config files
# Usage: get_mysql_option SECTION VARNAME DEFAULT
# result is returned in $result
# We use my_print_defaults which prints all options from multiple files,
# with the more specific ones later; hence take the last match.
get_mysql_option(){
	result=`/usr/bin/my_print_defaults "$1" | sed -n "s/^--$2=//p" | tail -n 1`
	if [ -z "$result" ]; then
	    # not found, use default
	    result="$3"
	fi
}

get_mysql_option mysqld datadir "/var/lib/mysql"
datadir="$result"
get_mysql_option mysqld socket "$datadir/mysql.sock"
socketfile="$result"
get_mysql_option mysqld_safe log-error "/var/log/mysqld.log"
errlogfile="$result"
get_mysql_option mysqld_safe pid-file "/var/run/mysqld/mysqld.pid"
mypidfile="$result"
get_mysql_option mysqld innodb_data_home_dir ""
innodb_data_home_dir="$result"

# if the host is fio, make sure the fio card is mounted
function check_fio_mount(){
  fio_in_dmesg=`dmesg | grep 'fioinf ioDrive' | wc -l`

  if [ "$fio_in_dmesg" != "0" ]; then
    has_flashcache=`ls /dev/* | grep -c '/dev/fioa[12]'`

    if [ "$has_flashcache" == '2' ]; then
      #TODO write a new function to detect if flashcache is enabled
      return 0
    fi

    mounted_fio=`mount | grep -c '/dev/fio'`
    if [ "$mounted_fio" == "0" ]; then
      return 1
    fi
  fi

  return 0
}

function check_mysql_procs(){
  if [ "`/usr/bin/pgrep mysqld_safe | wc -l`" == "0" ]; then
    return 0
  else
    return 1
  fi
}

start(){
	touch "$errlogfile"
	chown mysql:mysql "$errlogfile"
	chmod 0640 "$errlogfile"
	[ -x /sbin/restorecon ] && /sbin/restorecon "$errlogfile"

  # sanity check current setup
  if [ $force -ne 1 ]; then
    check_fio_mount
    if [ $? -ne 0 ]; then
      echo 'Fusion IO detected but not mounted (and not flashcache)'
      action $"Starting $prog: " /bin/false
      return 1
    fi
  	if [ ! -f '/etc/my.cnf' ]; then
      echo File not found: /etc/my.cnf
      action $"Starting $prog: " /bin/false
      return 1
    fi
    if [ ! -d "$datadir/mysql" ] ; then
        # we don't need to accidentally init the db here
        echo "MySQL database not found at $datadir/mysql"
        action $"Starting $prog: " /bin/false
        return 1
	  fi
    if [ ! -f "$innodb_data_home_dir/ibdata1" ]; then
      echo File not found: $innodb_data_home_dir/ibdata1
      action $"Starting $prog: " /bin/false
      return 1
    fi
    check_mysql_procs
    if [ $? -ne 0 ]; then
      echo mysqld process detected already running
      action $"Starting $prog: " /bin/false
      return 1
    fi
	fi
  chown mysql:mysql "$datadir"
	chmod 0755 "$datadir"
 
  # Pass all the options determined above, to ensure consistent behavior.
	# In many cases mysqld_safe would arrive at the same conclusions anyway
	# but we need to be sure.
	/usr/bin/mysqld_safe   --datadir="$datadir" --socket="$socketfile" \
		--log-error="$errlogfile" --pid-file="$mypidfile" \
		--user=mysql >/dev/null 2>&1 &
	ret=$?
  sleep 1
	# Spin for a maximum of N seconds waiting for the server to come up.
	# Rather than assuming we know a valid username, accept an "access
	# denied" response as meaning the server is functioning.
	if [ $ret -eq 0 ]; then
	    STARTTIMEOUT=60
	    while [ $STARTTIMEOUT -gt 0 ]; do
		RESPONSE=`/usr/bin/mysqladmin --socket="$socketfile" --user=UNKNOWN_MYSQL_USER ping 2>&1` && break
		echo "$RESPONSE" | grep -q "Access denied for user" && break
    
    sleep 1
    # if mysqld is not running, trigger immediate failure
    check_mysql_procs
    if [ $? -eq 0 ]; then
      action $"Starting $prog: " /bin/false
      return 1
    fi

		let STARTTIMEOUT=${STARTTIMEOUT}-1
	    done
	    if [ $STARTTIMEOUT -eq 0 ]; then
                    echo "Timeout error occurred trying to start MySQL Daemon."
                    action $"Starting $prog: " /bin/false
                    ret=1
            else
                    action $"Starting $prog: " /bin/true
            fi
	else
	    action $"Starting $prog: " /bin/false
	fi
	[ $ret -eq 0 ] && touch /var/lock/subsys/mysqld
	return $ret
}

stop(){
        MYSQLPID=`cat "$mypidfile"  2>/dev/null `
        if [ -n "$MYSQLPID" ]; then
            /bin/kill "$MYSQLPID" >/dev/null 2>&1
            ret=$?
            if [ $ret -eq 0 ]; then
                STOPTIMEOUT=600
                while [ $STOPTIMEOUT -gt 0 ]; do
                    /bin/kill -0 "$MYSQLPID" >/dev/null 2>&1 || break
                    sleep 1
                    let STOPTIMEOUT=${STOPTIMEOUT}-1
                done
                if [ $STOPTIMEOUT -eq 0 ]; then
                    echo "Timeout error occurred trying to stop MySQL Daemon."
                    ret=1
                    action $"Stopping $prog: " /bin/false
                else
                    rm -f /var/lock/subsys/mysqld
                    rm -f "$socketfile"
                    action $"Stopping $prog: " /bin/true
                fi
            else
                action $"Stopping $prog: " /bin/false
            fi
        else
            ret=1
            action $"Stopping $prog: " /bin/false
        fi
        return $ret
}

restart(){
    stop
    start
}

condrestart(){
    [ -e /var/lock/subsys/mysqld ] && restart || :
}

# See how we were called.
case "$1" in
  start)
    start
    ;;
  stop)
    stop
    ;;
  status)
    status mysqld
    ;;
  restart)
    restart
    ;;
  condrestart)
    condrestart
    ;;
  *)
    echo $"Usage: $0 {start|stop|status|condrestart|restart}"
    exit 1
esac

exit $?
